packages/app/components/EditMode/UploadButton/decorators/withUploadShapefile.ts

```diff
@@ -0,0 +1,54 @@
+import { compose, withHandlers } from 'recompose';
+import * as shapefile from 'shapefile';
+
+import withEditMode, {
+  WithEditModeProps
+} from '../../../decorators/withEditMode';
+
+export interface WithUploadShapefileProps {
+  handleUploadShapefile: (event: React.ChangeEvent<HTMLInputElement>) => void;
+}
+
+/**
+ * Read an uploaded file via <input type="file" /> as an ArrayBuffer.
+ *
+ * @param inputFile - The uploaded file.
+ * @see https://blog.shovonhasan.com/using-promises-with-filereader/
+ */
+const readAsArrayBuffer = (inputFile: File): Promise<ArrayBuffer> => {
+  const reader = new FileReader();
+  return new Promise((resolve, reject) => {
+    reader.onerror = () => {
+      reader.abort();
+      reject(new Error('Problem parsing shapeefile.'));
+    };
+
+    reader.onload = () => {
+      resolve(reader.result as ArrayBuffer);
+    };
+    reader.readAsArrayBuffer(inputFile);
+  });
+};
+
+/**
+ * Decorator to add a handler when we update the polygon we draw on the map.
+ */
+export default compose(
+  withEditMode,
+  withHandlers<WithEditModeProps, WithUploadShapefileProps>({
+    handleUploadShapefile: ({ updateGeojson }) => async ({
+      target: { files }
+    }: React.ChangeEvent<HTMLInputElement>) => {
+      try {
+        const uploaded = files[0];
```

Comment:
Maybe some validation should be performed here? A limit on the uploaded file size might be able to be configured on the web server, if not it should be done here. I added some client side validation to accept only .shp files, but there really should be some validation performed here as well to reject the request before we attempt to process it.